Dyno: make `VarScopeVisitor` analyses work per-tuple element

arifthpe opened the pull request:
Adjust `VarScopeVisitor` and the analyses implemented using it (call-init-deinit, copy elision, split init) to operate per-tuple element, as appropriate.

------------

## Explanation of approach

The main strategy here is to treat each tuple component within a tuple destructuring decl/assign as though it was a decl/assign of an individual variable, e.g., `var (a, (b,)) : (int, (real,)) = (1, (2.0,));` will be treated similarly to `var a : int = 1; var b : real = 2.0;`. VSV is adjusted to gather information on tuple decls/assigns as it goes (including nested tuples), but not holistically process them as decls/assigns. That processing only happens once each individual `VarLikeDecl`/`Identifier` contained within is reached, using info propagated from surrounding tuples to provide a pseudo- init-part and type-part as though it were a standalone decl/assign. As part of this approach, decl/assign handlers in VSV subclasses are refactored to accept a separate LHS and RHS rather than an actual decl/assign AST node, so they are agnostic to actual standalone decl/assigns vs the pseudo ones presented for destructured tuple elements.

`AssociatedAction`s are now generated for each tuple element as appropriate, and can be different for each element. It also includes additions to `AssociatedAction` itself, to accommodate cases where there is an Action on the tuple as a whole as well as its components:
- A tuple element index field, to indicate which tuple element the Action applies to when there is no actual AST for each element.
- A sub-actions field for Actions which contain others in a hierarchy. This is currently only used for tuples, but could also apply to other hierarchical Actions in the future if we add any.

An example where the above changes are relevant is the following:
```
record R { }
var r = new R();
var tup = (1, r);
var x = tup;
tup;
```

Here we would record an `INIT_OTHER` for `tup` which contained an `ASSIGN` and `MOVE_INIT` sub-action for its elements, respectively. Then we would have a `COPY_INIT` for `x` which contained an `ASSIGN` and `COPY_INIT` for its elements, respectively, and since there is no RHS AST to store the IDs of in the sub-actions to keep track of which action belongs to which element, we would store the element indices.

A major drawback of this approach is that it does not elegantly handle the case of a (non-destructuring) tuple _variable_ LHS and a tuple _expression_ RHS, like `var myTuple = (a, b);`. The VSV changes are centered around traversing the LHS and creating an equivalent decl/assign for each LHS component, but in this case there is no AST to traverse that represents the tuple elements. However, some VSV analyses still need to handle a tuple expr RHS on a per-component basis, such as noting variable mentions and finding elided copies. This PR does so by adding special behavior to each of the VSV subclasses which need to do such handling. Some alternatives considered were:
- Resolving a generated `init=`/`=` call on the tuple which would perform per-element initialization/assignment. This should give us handling of copy elision etc for free, as each RHS component would be an argument to the call and processed as those normally are. I'm not sure if this approach could also cover the destructuring decl/assign case well. I didn't look into this route because I didn't think of it until Ben suggested it when I asked for input, by which time I had already nearly finished the VSV subclass special behavior approach.
  - Upon further investigation of using this method to handle the tuple var = tuple expr case, I still decided against it for the time being. Generating `init=`/`=` for the tuple would require changing our function generation infrastructure from making unique functions per-type-id to per-instantiation, as all tuples are instantiations of `ChapelTuple._tuple`. Additionally, we could just use the existing `_tuple.=` and a new similar `_tuple.init=` which assigns each corresponding element pair in a param for loop, but VSV analyses don't handle param for loops yet (https://github.com/Cray/chapel-private/issues/4209). So I think it makes the most sense to keep the special behavior for this case for the time being, then rip it out and use module code calls once VSV can properly handle that.
- Traversing the RHS of such decl/assigns in VSV, and generating a pseudo standalone decl/assign for each LHS/RHS component pair. Since there is no LHS component AST it wouldn't fit perfectly into the approach of making VSV analyses agnostic to tuple component init/assigns, and likely require some more adjustments to each analysis. I partly implemented then abandoned this approach due to its excessive complexity.
- Some other approach which unifies both the destructuring and non-destructuring decl/assign cases, so both can be presented as pseudo-standalone decl/assigns per-element to the VSV analyses. I was unable to come up with a way to do this that wasn't way more complicated than it's worth.

This PR also reverts the change from https://github.com/chapel-lang/chapel/pull/27533 which changed tuple expressions containing any refs from `CONST_VAR` to `CONST_REF` kind, as discussed from https://github.com/chapel-lang/chapel/pull/27533#issuecomment-3130229056.

------------

Contributes to https://github.com/Cray/chapel-private/issues/6283.

Future work:
- Add `DEINIT` for tuples (part of https://github.com/Cray/chapel-private/issues/6283). Not included due to the size of this PR.

Note to reviewer: The commit history of this branch is a mess due to abandoning an earlier approach, so it'd be best to review the diff as a whole.

[reviewed by @DanilaFe , thanks!]

Testing:
- [x] dyno tests
- [x] paratest

arifthpe committed:
Add tests for tuple unpacking variable init/assignment

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Test normal assignment rather than split-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Switch tuple expression kind back to CONST_VAR

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Get copy init for tuple unpack assignment working

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Comment out most CID tests for iterating on tuples

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add stub VarScopeVisitor::handleTupleDeclaration

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Stub out resolveTupleInit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Uncomment more tuple tests

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add sub-actions and second ID to AssociatedAction

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Set up testing sub-actions and other ID in testCallInitDeinit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add some tests for tuple init= with sub actions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Reorder and stringify new AssociatedAction fields

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Restore semantics of getLhsForTupleUnpackAssign

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Resolve init for each tuple elt

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Resolve sub-actions per tuple idx

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add AssociatedAction::dump

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix sub-action hierarchy for nested tuple expr

Plus code and comment cleanups

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Adjust tuple destructuring LHS types to var

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use type alias for sub-actions field

Plus adjust AssociatedAction::mark for older compiler used in CI

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
wip checkpoint on copy elision/split init for destructuring assignment

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Rename resolveTupleUnpackAssignOrInit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Cleanups

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Checkpoint on destructuring declarations

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix actions generated for destructuring decl

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Uncomment tests

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Implement tuple destructuring decl copy elision

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Implement tuple destucturing decl copy elision

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix saving multiple elision points for tuple destructuring decl

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix copy elision for tuple destructuring assignment

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Allow TupleDecl to contain Formals

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix some attempts to use nonexistent init expr

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove straggling breakpoint

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Improve test cases

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Skip underscores in tuple decl copy elision

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Implement copy elision from tuple expr RHS

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Avoid move init for tuple elts (rather than assign)

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Expect VAR kind for lhs tuple expr elts

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Expect tuple = actions on op rather than elt

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Emit tuple unpack assign size mismatch error just once

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add test for tuple expr containing tuple vars

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix handling of tuple expr containing tuple vars

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Skip default-init of tuple-grouped loop indices

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove debugger breakpoints

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use helper for loop index decl check, make VSV enter tuples

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Track current position in tuple decl nests

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Clean up and document VSV helper methods

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix some index mistakes

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Unify decl handling interface between VSV subclasses

Only set up for individual decls, will need to adjust to support decls
in tuple decls

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use tuple decl's properties for nested decls

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix bad autoformat

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix AST stack indexing and handle unknown types

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix recovering decl type info from containing tuple decl

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
call-init-deinit add tuple idx to associated actions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix copy elision from tuple var

Includes maintaining a stack of nested tuple exprs used for tuple init
part, alongside the type stack.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Skip _ var decls in VarScopeVisitor

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Loop through tuple expr RHS in copy elision

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix copy elision on nested tuple expr

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Expect MOVE_INIT after NEW_INIT as appropriate

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Additional tuple copy elision fixes, simplify CID

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add more tuple split init test cases

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
checkpoint

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix traversal of tuple expr assignments

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix order of traversing nested tuples, plus refactor

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Adjust tuple expr special case comment

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Pass rhsAst to more processInit call sites

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Skip _ in tuple assign

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Put back tuple unpack assignment kind adjustments

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix detecting split-init via tuple destructuring

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Set rhsAst for tuple assign even if RHS isn't tuple expr

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Count copy elision in assignment LHS rather than assignment itself

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Track copy-elided tuple expr RHS components

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Properly handle AssociativeAction changes in testCallInitDeinit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use __FUNCTION__ instead of str literals in testCallInitDeinit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix test numbering order

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix size_t to int comparison

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Address some TODOs

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove commented code from old approaches

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Comment tuple special behavior in addCopyInit

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix using marker fn for subactions SmallVector

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove unused CallInitDeinit:validateTypeAndInitExpr

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use constant instead of hardcoded -1 for no tuple elt

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Make a handful of minor code and comment revisions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Combine paired tuple info vectors into vector of pairs

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -187,7 +194,26 @@ static bool isTypeParam(QualifiedType::Kind kind) {
                                           RV& rv) {
-  return rv.byAst(rhsAst).toId().isEmpty() && !isRef(rhsType.kind());
+  if (rhsAst->isTuple() && rhsType.type() && rhsType.type()->isTupleType()) {
+    if (auto asTupleType = rhsType.type()->toTupleType()) {
```
How can `rhsAst->isTuple()` be true and not `rhsType.type()->isTupleType()`? Are you guarding against `Unknown` or `Erroneous`? If so, consider making that explicit. Otherwise, consider an assertion instead.

DanilaFe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -882,41 +808,86 @@ void CallInitDeinit::processInit(VarFrame* frame,
+          auto rhsTupleExpr = rhsAst->toTuple();
+
+          auto& re = rv.byAst(ast);
+          auto topLevelAction = re.associatedActions().front();
```
probably need an assert here that there is an associated action. This wasn't even true prior to recently (e.g., `var x: int; var y = x;` probably was called here but `MOVE_INIT` is very recent).

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

arifthpe committed:
Remove special handling of tuple var = tuple expr case

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Rely on module code init= for tuple init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Revert "Rely on module code init= for tuple init"

This reverts commit 4170bba0ca2603ddb5c289e1b03cbf269351ba42.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Revert "Remove special handling of tuple var = tuple expr case"

This reverts commit ac9fae0ad05ff0194fed8da684b015b279f1f992.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Comment special behavior for tuple expr RHS

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe reviewed (commented):
One batch of comments

arifthpe committed:
Fix incorrect invocation of processInit tuple expr RHS case

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Move processInit tuple expr RHS logic into helper

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add assertions about topLevelAction existence and kind

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Refactor copy-elided init check into helper

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -846,30 +770,107 @@ void CallInitDeinit::resolveMoveInit(const AstNode* ast,
+      auto useTupleEltIdx = i;
+      if (actual) {
+        useId = actual->id();
+        useTupleEltIdx = AssociatedAction::NO_TUPLE_ELT;
```
Huh... why does having a corresponding actual reset index to `NO_TUPLE_ELT`?

DanilaFe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -968,18 +983,16 @@ void CallInitDeinit::resolveDeinit(const AstNode* ast,
-  // check for use of deinited variables in type or init exprs
-  if (auto type = ast->typeExpression()) {
-    processMentions(type, rv);
-  }
```
I believe you that this is safe to remove, but I am curious if you have an intuition for what it was used for originally and why it isn't needed anymore.

DanilaFe commented on frontend/lib/resolution/copy-elision.cpp:
```diff
@@ -266,62 +288,139 @@ void FindElidedCopies::noteMentionsForOutFormals(VarFrame* frame) {
+    QualifiedType lhsType = rv.byId(lhsVarId).type();
+    auto tupleExprInit = rhsAst->toTuple();
+    if (lhsType.type() && lhsType.type()->isTupleType() &&
+        tupleExprInit) {
```
Should that TODO/HACK comment reoccur here?

arifthpe replied:
This logic is since removed

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

arifthpe committed:
Restore size template param to SmallVector mark

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use optional<int> instead of sentinel for AssociatedAction::tupleEltIdx

Also rename accessor from tupleEltIndex to tupleEltIdx, for consistency
with sibling accessors.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Avoid copying AssociatedActions and its constructor args

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Create ElidedCopyInfo typedef for copy-elision results

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use const auto& for iteration

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix error message printing for sub-actions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Store tuple elt idx even when elt ID exists

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Revert "Restore size template param to SmallVector mark"

This reverts commit b477076ba8319dab1f8b31f2d7b99a011ad58b4c.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Make AssociatedAction own and destroy sub-actions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Switch to unique_ptr, avoid some AA copies

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Add and use OpCall::{lhs,rhs} helpers

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Expect non-null RHS in var = expr case

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe commented on frontend/include/chpl/resolution/resolution-types.h:
```diff
@@ -2587,26 +2587,85 @@ class AssociatedAction {
+  // Currently only used for tuple call-init-deinit, where each element may
+  // have a sub-action.
+  // Sub-actions are owned by the parent, and will be destroyed with it.
+  ActionsList subActions_;
```
Remind me, is there a reason why these are heap-allocated externally instead of stored by value in the `ActionsList`? Is it because you do a lot of action-moving? If so, I think you can get away with `.swap()` on `ActionList`s instead of element-by-element copy.

arifthpe replied:
I think this comment was made when `ActionsList` was an `llvm::SmallVector` of `AssociatedAction*` instead of a `std::vector` of `AssociatedAction`. And the reason I had stored `AssociatedAction*`s was because I didn't realize the use of `SmallVector` (which may put elements on the stack) was the reason the compiler was balking at storing `AssociatedAction`s within `AssociatedAction`, whereas `vector` puts the values in the free store.

arifthpe committed:
Add another use of OpCall::lhs

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe commented on frontend/lib/resolution/Resolver.cpp:
```diff
@@ -3268,9 +3268,20 @@ bool Resolver::resolveSpecialOpCall(const Call* call) {
+
+      if (auto lhsTuple = op->lhs()->toTuple()) {
+        auto rhsQt = byPostorder.byAst(op->rhs()).type();
+        if (auto rhsTupleType = rhsQt.type()->toTupleType()) {
```
What mechanism prevents this error from being issued in the "normal" way?

arifthpe replied:
As in an error class I assume? I think it was laziness. Modified `TupleDeclMismatchedElems` to be for decl or assign and used that.

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -314,25 +477,54 @@ void VarScopeVisitor::exit(const NamedDecl* ast, RV& rv) {
+
+        handleDeclaration(vld, parent, initExpr, initType, intentOrKind, isFormal,
+                        rv);
+      }
```
Can we put the tuple helping code into a sub-proc and maybe re-arrange this so that the more common, non-tuple case comes first?

arifthpe replied:
I think moving the tuple branch into a helper makes the intent less clear, as it hides the symmetry between the tuple and non-tuple cases. In the tuple branch we gather the decl properties from the surrounding tuple decl, and in the non-tuple branch we gather these same properties directly from the decl, then call `handleDeclaration` in both cases.

I am on board with moving the non-tuple case to come first, I'll do that.

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Revert "Switch to unique_ptr, avoid some AA copies"

This reverts commit 3f1677c96338dc72d04312604aa494c3345efdbb.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Revert "Make AssociatedAction own and destroy sub-actions"

This reverts commit 3965ef4b5bd9ba9f1438b0c745f7e20cfc7296b8.

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use std::vector for subActions list

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Restore dead vars check CallInitDeinit::handleDeclaration

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove support for per-element copy elision from tuples

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -187,7 +203,29 @@ static bool isTypeParam(QualifiedType::Kind kind) {
+    if (parentAst->isTuple() || parentAst->isTupleDecl()) {
+      isProbablyCall = false;
+    }
+  }
```
I tested with the production compiler with `--log` and found after desugaring to that tmp form it has an extra copy (maybe two, via passing arg to `_build_tuple` plus copy-init of `x`). However, after the scalarReplace pass the tuple goes away entirely, and then after ref and copy propagation and DCE there are no extra copies happening AFAICT.

I am not really sure if this means dyno ought to output this as a copy or not. Since those later steps which get rid of the tuple will happen after dyno anyways, I don't think it makes a difference in compilation output, and it is faithful to how production treats tuples until they get stripped out entirely. So IMO it's fine for this PR, but do you think that warrants a TODO/future work note to avoid the copy in Dyno?

<details>
<summary>Test file and log outputs</summary>

Test file:
```chapel
record R {
}

proc main() {
  writeln(one());
  writeln(two());
}

proc one() {
  var (x,) = (new R(),);
  return x;
}
proc two() {
  var x = new R();
  return x;
}
```

Snippet of `log/mytest_12resolve.ast`:
```
  function one[395422]() : R[395407] "resolved"
  {
    unknown ret[668690]:R[395407] "dead after last mention" "RVV" "temp"
    unknown call_tmp[1039145]:R[395407] "dead after last mention" "insert auto destroy for explicit new" "maybe param" "temp"
    val initTemp[2021543]:R[395407] "dead after last mention" "temp"
    (2021546 call init[1485820] initTemp[2021543])
    (1039148 'move' call_tmp[1039145] initTemp[2021543])
    unknown call_tmp[1039150]:1*R[2021636] "dead after last mention" "maybe param" "temp"
    (1039153 'move' call_tmp[1039150](395432 call _build_tuple[2021602] call_tmp[1039145]))
    unknown tmp[395434]:1*R[2021765] "dead at end of block" "insert auto destroy" "temp"
    (684671 'move' tmp[395434](2021700 call chpl__initCopy[2021703] call_tmp[1039150] 0))
    (603235 'end of statement')
    ref call_tmp[1039155]:_ref(R)[2021526] "dead after last mention" "insert auto destroy" "maybe param" "temp"
    (1039158 'move' call_tmp[1039155](2022080 '.' tmp[395434] x0[2021763]))
    unknown x[395426]:R[395407] "dead at end of block" "insert auto destroy"
    (684673 call init=[1485882] x[395426] call_tmp[1039155])
    (395443 'end of statement')
    (668695 'move' ret[668690] x[395426])
    gotoReturn _end_one[668688] _end_one[668688]
    label _end_one[668688]
    (668692 return ret[668690])
  }
  function two[395448]() : R[395407] "resolved"
  {
    unknown ret[668701]:R[395407] "dead after last mention" "RVV" "temp"
    unknown call_tmp[1039160]:R[395407] "dead at end of block" "maybe param" "temp"
    val initTemp[2031983]:R[395407] "dead at end of block" "temp"
    (2031986 call init[1485820] initTemp[2031983])
    (1039163 'move' call_tmp[1039160] initTemp[2031983])
    unknown x[395452]:R[395407] "dead at end of block" "insert auto destroy"
    (684675 'move' x[395452] call_tmp[1039160])
    (395457 'end of statement')
    (668706 'move' ret[668701] x[395452])
    gotoReturn _end_two[668699] _end_two[668699]
    label _end_two[668699]
    (668703 return ret[668701])
  }
```

Snippet of `log/mytest_26scalarReplace.ast`:
```
  function one[395422](ref arg _retArg[2445087]:R[395407]) : void[4] "fn returns via _retArg" "resolved"
  {
    const-val ret[668690]:R[395407] "dead after last mention" "RVV" "temp"
    const-val call_tmp[1039145]:R[395407] "dead after last mention" "insert auto destroy for explicit new" "maybe param" "temp"
    const-val initTemp[2021543]:R[395407] "dead after last mention" "temp"
    (2021546 call init[1485820] initTemp[2021543])
    (1039148 'move' call_tmp[1039145] initTemp[2021543])
    const-ref call_tmp_x0[2846684]:R[395407] "temp"
    const-ref ret_tmp_x0[2846692]:R[395407] "temp"
    ref i__e0_x[2810710]:R[395407] "temp"
    (2810714 'move' i__e0_x[2810710](2810712 'set reference' call_tmp[1039145]))
    const-ref this_x0[2846700]:R[395407] "temp"
    (2810725 'move' this_x0[2846700] i__e0_x[2810710])
    (2846698 'move' ret_tmp_x0[2846692] this_x0[2846700])
    (2846690 'move' call_tmp_x0[2846684] ret_tmp_x0[2846692])
    val tmp_x0[2846716]:R[395407] "temp"
    val inlineImm[2810743]:bool[19] "temp"
    (2810746 'move' inlineImm[2810743] 0)
    val retv_x0[2846725]:R[395407] "temp"
    const-ref read_x0[2810751]:R[395407] "dead after last mention" "temp"
    (2810753 'move' read_x0[2810751] call_tmp_x0[2846684])
    val elt_x0[2810758]:R[395407] "dead after last mention" "temp"
    (2810760 'move' elt_x0[2810758] read_x0[2810751])
    (2810763 'move' retv_x0[2846725] elt_x0[2810758])
    (2846722 'move' tmp_x0[2846716] retv_x0[2846725])
    ref i_x[2827111]:R[395407] "temp"
    (2827115 'move' i_x[2827111](2827113 'set reference' call_tmp[1039145]))
    const-ref call_tmp[1039155]:R[395407] "dead after last mention" "insert auto destroy" "maybe param" "temp"
    (1039158 'move' call_tmp[1039155](2846724 'set reference' tmp_x0[2846716]))
    const-val x[395426]:R[395407] "dead at end of block" "insert auto destroy"
    (684673 call init=[1485882] x[395426] call_tmp[1039155])
    (668695 'move' ret[668690] x[395426])
    label _end_one[668688]
    (2445089 '=' _retArg[2445087] ret[668690])
    (2445091 return _void[46])
  }
  function two[395448](ref arg _retArg[2445099]:R[395407]) : void[4] "fn returns via _retArg" "resolved"
  {
    const-val ret[668701]:R[395407] "dead after last mention" "RVV" "temp"
    const-val call_tmp[1039160]:R[395407] "dead at end of block" "maybe param" "temp"
    const-val initTemp[2031983]:R[395407] "dead at end of block" "temp"
    (2031986 call init[1485820] initTemp[2031983])
    (1039163 'move' call_tmp[1039160] initTemp[2031983])
    const-val x[395452]:R[395407] "dead at end of block" "insert auto destroy"
    (684675 'move' x[395452] call_tmp[1039160])
    (668706 'move' ret[668701] x[395452])
    label _end_two[668699]
    (2445101 '=' _retArg[2445099] ret[668701])
    (2445103 return _void[46])
  }
```

Snippet of `log/mytest_29deadCodeElimination.ast`:
```
  function one[395422](ref arg _retArg[2445087]:R[395407]) : void[4] "fn returns via _retArg" "resolved"
  {
    const-val ret[668690]:R[395407] "dead after last mention" "RVV" "temp"
    const-val call_tmp[1039145]:R[395407] "dead after last mention" "insert auto destroy for explicit new" "maybe param" "temp"
    const-val initTemp[2021543]:R[395407] "dead after last mention" "temp"
    (1039148 'move' call_tmp[1039145] initTemp[2021543])
    const-val x[395426]:R[395407] "dead at end of block" "insert auto destroy"
    (668695 'move' ret[668690] x[395426])
    label _end_one[668688]
    (2445089 '=' _retArg[2445087] ret[668690])
    (2445091 return _void[46])
  }
  function two[395448](ref arg _retArg[2445099]:R[395407]) : void[4] "fn returns via _retArg" "resolved"
  {
    const-val call_tmp[1039160]:R[395407] "dead at end of block" "maybe param" "temp"
    const-val initTemp[2031983]:R[395407] "dead at end of block" "temp"
    (1039163 'move' call_tmp[1039160] initTemp[2031983])
    label _end_two[668699]
    (2445101 '=' _retArg[2445099] call_tmp[1039160])
    (2445103 return _void[46])
  }
```

Snippet of `log/mytest_40codegen.ast`:
```
  function one[395422]:wide proc(ref _retArg: _ref(R))[2895055](ref arg _retArg[2445087]:_ref(R)[2021526]) : void[4] "fast on" "fn returns via _retArg" "local fn" "resolved"
  {
    const-val ret[668690]:R[395407] "dead after last mention" "RVV" "temp"
    const-val call_tmp[1039145]:R[395407] "dead after last mention" "insert auto destroy for explicit new" "maybe param" "temp"
    const-val initTemp[2021543]:R[395407] "dead after last mention" "temp"
    (1039148 'move' call_tmp[1039145] initTemp[2021543])
    const-val x[395426]:R[395407] "dead at end of block" "insert auto destroy"
    (668695 'move' ret[668690] x[395426])
    (2445089 '=' _retArg[2445087] ret[668690])
    (2445091 return _void[46])
  }
  function two[395448]:wide proc(ref _retArg: _ref(R))[2895055](ref arg _retArg[2445099]:_ref(R)[2021526]) : void[4] "fast on" "fn returns via _retArg" "local fn" "resolved"
  {
    const-val call_tmp[1039160]:R[395407] "dead at end of block" "maybe param" "temp"
    const-val initTemp[2031983]:R[395407] "dead at end of block" "temp"
    (1039163 'move' call_tmp[1039160] initTemp[2031983])
    (2445101 '=' _retArg[2445099] call_tmp[1039160])
    (2445103 return _void[46])
  }
```

</details>

arifthpe replied:
As far as the logic in this method, I moved handling the cases of the RHS being a tuple element, or the whole decl/assign being a tuple destructuring, up into `processInit` with a comment. Removed the incorrect description of being a tuple component as being tied to being a value or not.

arifthpe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -846,30 +770,107 @@ void CallInitDeinit::resolveMoveInit(const AstNode* ast,
+bool CallInitDeinit::initIsCopyElidedFrom(const AstNode* lhsAst,
+                                          const AstNode* rhsAst,
+                                          const ID& rhsDeclId) {
+  for (const ID& id : {lhsAst->id(), rhsAst->id()}) {
```
IIRC we figured out this was because I had _meant_ to change copy-elisions to be tracked by LHS, but had missed some cases in which it was tracked by the AST of the entire op (e.g., `LHS = RHS`). And in the case in `processInit` where we set `rhsAst = ast`, I was unwittingly relying on checking the RHS to handle those missed cases.

Anyways, with per-elt copy elision analysis for tuples reverted, the reason for making that change is gone, so we now always track by "the whole AST" as before.

arifthpe commented on frontend/lib/resolution/copy-elision.cpp:
```diff
@@ -266,62 +288,139 @@ void FindElidedCopies::noteMentionsForOutFormals(VarFrame* frame) {
+            lhsTupleType.type()->toTupleType()->elementType(i);
+        // Remove ref-ness from lhs tuple elt since we are using it as a var,
+        // unless the whole tuple itself is ref
+        if (!lhsTupleType.isRef()) {
```
I think this comment predated the move into a helper in https://github.com/chapel-lang/chapel/pull/27563/changes/343c81f1b7e312cf033b99e479cb8d7fc94300d2. There is now just the `TODO/HACK` comment here, logic to check if this is that special case, then a call to the helper. Let me know if that looks understandable, I could also move all the check logic into the helper and just unconditionally call the helper here.

arifthpe commented on frontend/lib/resolution/copy-elision.cpp:
```diff
@@ -266,62 +288,139 @@ void FindElidedCopies::noteMentionsForOutFormals(VarFrame* frame) {
+      QualifiedType lhsTupleType = initType;
+      CHPL_ASSERT(lhsTupleType.type() &&
+                  lhsTupleType.type()->isTupleType());
+      for (int i = 0; i < tupleExprInit->numActuals(); i++) {
```
Yep, removed

arifthpe commented on frontend/lib/resolution/copy-elision.cpp:
```diff
@@ -266,62 +288,139 @@ void FindElidedCopies::noteMentionsForOutFormals(VarFrame* frame) {
+            // check that the types are the same
+            if (rv.hasId(lhsVarId) && rv.hasId(rhsVarId)) {
+              if (lhsType.type() && lhsType.type()->isTupleType()) {
+                const TupleType* ttype = lhsType.type()->toTupleType();
```
Moot now as this code is removed (but yes)

arifthpe committed:
Put VSV NamedDecl not-within-tuple case first

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Print each test name in testLoopIndexVars

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Use computed type instead of type expr for TupleDecl

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -282,18 +369,94 @@ bool VarScopeVisitor::enter(const TupleDecl* ast, RV& rv) {
+  // top-level decl, or from the parent tuple decl in the case of nesting.
+  QualifiedType initType;
+  if (auto typeExpr = ast->typeExpression()) {
+    initType = rv.byAst(typeExpr).type();
```
I think we can assume it's been computed already since VSV analyses run after resolution. Switched

arifthpe commented on frontend/lib/resolution/copy-elision.cpp:
```diff
@@ -230,7 +248,9 @@ void FindElidedCopies::saveElidedCopies(VarFrame* frame) {
     if (state.lastIsCopy) {
-      allElidedCopyFromIds.insert(state.points.begin(), state.points.end());
+      for (auto point : state.points) {
+        allElidedCopyFromIds.emplace(point, pair.first);
```
Per offline discussion and some other comments, yes this was inconsistent and has been reverted

arifthpe committed:
Emit proper error for mismatched size tuple assign

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -282,18 +369,94 @@ bool VarScopeVisitor::enter(const TupleDecl* ast, RV& rv) {
+    auto decl = ast->decl(i);
+    auto& re = rv.byPostorder().byAst(decl);
+    llvm::SmallVector<AssociatedAction> subActions;
+    for (auto& action : re.associatedActions()) {
```
My intention with putting the AssociatedActions into their own (essentially) tree attached to the top-level decl/assign, rather than just putting each one in its appropriate place in the AST itself, was to be able to represent the AssociatedActions in every case of `{tuple expr LHS, tuple var LHS} x {tuple expr RHS, tuple var RHS}` in a uniform way. If we wanted to keep track of separate actions per-element for a tuple var LHS, there is no sub-decl AST to attach an action to, so we have to put them on the parent, and for consistency I just always did that.

arifthpe committed:
Add nested destructuring decl test

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Move some logic for forcing tuple copy out of isCallProducingValue

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -187,7 +203,29 @@ static bool isTypeParam(QualifiedType::Kind kind) {
+    return asTupleType == asTupleType->toValueTuple(context);
+  }
+
+  bool isProbablyCall = rv.byAst(rhsAst).toId().isEmpty();
```
Inlined this variable (as it was pre-PR) as part of moving the tuple-specific logic out

arifthpe committed:
Improve logic for detecting RHS tuple expr

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Condense forced tuple copy logic to a single bool

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Handle RHS tuple expr containing tuple expansions

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -187,7 +203,29 @@ static bool isTypeParam(QualifiedType::Kind kind) {
+  // a false positive.
+  auto parentId = parsing::idToParentId(context, rhsAst->id());
+  if (auto parentAst = parsing::idToAst(context, parentId)) {
+    if (parentAst->isTuple() || parentAst->isTupleDecl()) {
```
Decided to force copy in both these cases, but made the check for them explicit (and cleaner), commented, and moved it into `processInit`

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Fix iterator variable used for elt idx

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Support hashing of chpl::optional

Signed-off-by: Anna Rift <anna.rift@hpe.com>

DanilaFe commented on frontend/include/chpl/resolution/copy-elision.h:
```diff
@@ -24,10 +24,12 @@
 #include "chpl/resolution/resolution-types.h"
 
 #include <set>
+#include <unordered_map>
```
unused now?

DanilaFe commented on frontend/include/chpl/resolution/resolution-types.h:
```diff
@@ -2651,15 +2670,33 @@ class AssociatedAction {
 
   const types::QualifiedType type() const { return type_; }
 
+  const chpl::optional<int> tupleEltIdx() const {
```
I suspect you meant `const chpl::optional<int>&`.

DanilaFe commented on frontend/lib/resolution/call-init-deinit.cpp:
```diff
@@ -846,32 +756,107 @@ void CallInitDeinit::resolveMoveInit(const AstNode* ast,
+                                /* id */ actual->id(),
+                                action.type(),
+                                /* tupleEltIdx */ eltIdx,
+                                action.subActions());
```
Should you be `std::move`ing here?

DanilaFe commented on frontend/lib/resolution/Resolver.cpp:
```diff
@@ -3276,9 +3276,19 @@ bool Resolver::resolveSpecialOpCall(const Call* call) {
 
   if (op->op() == USTR("=")) {
-    if (op->numActuals() == 2) {
+    if (op->isBinaryOp()) {
```
This probably ought to be an assertion rather than a second condition, right?

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -176,15 +176,102 @@ VarScopeVisitor::processSplitInitOut(const Call* ast,
+
+// Adjusts LHS tuple type so that its components are all values.
+// Does no sanity checks.
+static QualifiedType getLhsForTupleUnpackAssign(Context* context,
```
Should / can we just use `getValueTuple`?

arifthpe replied:
I tried it out, and it fails because `getLhsForTupleUnpackAssign` sets the type of `_` elements to `NothingType`, which `getValueTuple` and `toValueTuple` do not. We could relax the requirement that `_` resolves to `NothingType` (since I think everywhere that processes tuples skips `_` anyways), or otherwise just go through and resolve any `_` actuals to `NothingType` right here and then hand off to `getValueTuple`. I'm leaning towards the latter option, though it has the downside of an additional loop through the tuple's actuals.

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -282,18 +369,86 @@ bool VarScopeVisitor::enter(const TupleDecl* ast, RV& rv) {
+                              action.subActions());
+    }
+
+    re.clearAssociatedActions();
+    for (auto action : subActions) {
+      re.addAssociatedAction(std::move(action));
+    }
```
Here, this confuses me. For `var x = (a, b, c)` I can see storing subactions for `a`, `b`, and `c`. But for `var (x, y, z) = (a, b, c)`, why is it not fine to leave subactions on the sub-decls? Consistency?

arifthpe replied:
Yes, just for consistency https://github.com/chapel-lang/chapel/pull/27563#discussion_r3095740232

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -314,25 +469,62 @@ void VarScopeVisitor::exit(const NamedDecl* ast, RV& rv) {
+        } else {
+          parent = parsing::parentAst(context, outerTupleDecl);
+          initExpr = outerTupleDecl->initExpression();
+          auto parentInitExpr = nestedTupleInfoStack.back().second;
```
Is the goal of this to handle:

```C++
var (x, y, z) = (A, B, C); // (1)
var (x, y, z) = a; // (2)
```

?

Where, for `x` you pick `A` if available, and if not, pick a more general `a`? 

If so, I'm not 100% sure on this. The 'this-or-outermost' approach was a bit confusing to me. What about:

```
var ((x,), y, z) = (A, B, C)
```

Here, the single-element tuple `A` is being unpacked, so this is kind of a mix of `(2)` and `(1)` above. But this logic would pick `(A, B, C)` as the init expression for `x`, no?

arifthpe replied:
Yes, that's the intention, to store the logical corresponding init expr on a per-element basis if possible, and otherwise just the whole tuple's init expr if there is no such corresponding component for a given elt. It is kind of weird and I could see an argument for just using `nullptr` as the init expr for an element if we don't have an actual init expr to show for it. I'm not sure off the top of my head if that would break anything relying on having the outermost init expr.

In the second example you gave, assuming `A` is a tuple variable and not standing in for a single-elt tuple expr, I would have expected `A` to be the init expr for `x`, but I tested and it was `(A, B, C)`. This happens because `A` is not a tuple expr and the `enter(Tuple)` handler will only store an init-part if it's a tuple expr. I don't think there's any reason we couldn't adjust that so `A` gets stored as the init-part (adjusting uses that assume it's a tuple and try to index into it), and then `A` would be used as the init-part for `x`. I think that'd be more consistent with the non-nested case and I'd like to do it, but will wait to hear back on if you think we should abandon the "this-or-outermost" approach entirely first.

DanilaFe replied:
Yes, `A` is meant to be a tuple variable, not a literal. 

Personally, I think "this-or-outermost" is quite weird. "Outermost" has little to do with the specific (deeply nested, potentially) tuple de-structuring. I think I'd be okay with this if the "logical corresponding init expr" were just the "closest single expression in a matched traversal", so that we store `a` in the original case, `A` in the singleton case (A is the closest matching sub-expression for `x`), and so on.

I don't know how this fits into your current implementation and if your proposed approach would do this. I think we ought to have `A` as the `init-expr` for all the tuples here:

```chapel
var ((((a,),),),) = A
```

Your suggestion would be to set `A` as the `init-expr` for each of the corresponding entries on the tuple-stack, right? I think that would be fine (and would fall naturally out of your existing logic, I think).

arifthpe replied:
I guess in my mind I was only thinking about propagating a non-tuple-expr init-part one level deep, so in the `var ((x,), y, z) = (A, B, C)` example `A` would be the init-expr for both `(x,)` and `x`, and I didn't think about nesting further than that -- I was assuming deeper init-parts would still be `nullptr`. But I don't know how I would actually implement going just one layer "further", nor would it really make sense to treat that specific level of nesting specially. So I'm on board with propagating `A` all the way down.

And yes, it falls out of the current logic very easily. Instead of "this elt's init-part is the corresponding component of the parent if parent is a tuple, or `nullptr` otherwise", just use the parent instead of `nullptr`.

I've implemented this now.

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -345,13 +537,25 @@ bool VarScopeVisitor::enter(const OpCall* ast, RV& rv) {
+      rhsAst->traverse(rv);
+
+      // Set containing tuple assignment only while traversing LHS
+      inTupleAssignment = ast;
+      lhsAst->traverse(rv);
+      inTupleAssignment = nullptr;
```
I haven't thought too deeply about this, but this pattern usually susses me out if we expect recursion and this function getting called in a nested fashion, since it clobbers previous `inTupleAssignment` state. Two ways to handle this:

1. Assert `inTupleAssignment` is null
2. `std::swap` `inTupleAssignment` into a temporary variable, then swap back.

arifthpe replied:
I will go with asserting `!isTupleAssignment`, since that is the (previously unwritten) assumption

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.cpp:
```diff
@@ -363,6 +567,96 @@ void VarScopeVisitor::exit(const OpCall* ast, RV& rv) {
+  auto parentAst = parsing::parentAst(context, ast);
+  if (parentAst == inTupleAssignment) {
+    CHPL_ASSERT(ast == inTupleAssignment->actual(0));
+    auto rhsAst = inTupleAssignment->actual(1);
```
`rhs()`?

DanilaFe commented on frontend/lib/resolution/VarScopeVisitor.h:
```diff
@@ -64,14 +64,31 @@ class VarScopeVisitor : public BranchSensitiveVisitor<VarFrame, MutatingResolved
+  // Matches the number and order of tuples/tuple decls in the AST stack.
+  llvm::SmallVector<std::pair<types::QualifiedType, const Tuple*>>
+      nestedTupleInfoStack;
+  // Tuple assignment we are currently within, if any.
```
should probably note that it's only if we're in the LHS (see your comments in the implementation above).

DanilaFe reviewed (commented):
I think only minor concerns remain. I ran out of juice before getting to the tests and hope to reprise that effort a little later today.

DanilaFe reviewed (approved):
Tests are simpler than I thought. Barring other suggestions I submitted, looks good!

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

arifthpe committed:
Remove unused unordered_map include

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Apply more reviewer feedback

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Eliminate getLhsForTupleUnpackAssign helper

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Clarify inTupleAssignment set only within LHS

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Rename some local variables

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Preserve non-tuple-expr 'closest' init parts

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Remove stray debugger breakpoint

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Merge branch 'main' into tuple-unpacking-copy-init

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Avoid asserting on invalid tuple destructuring LHS

Signed-off-by: Anna Rift <anna.rift@hpe.com>

arifthpe committed:
Update tuple size mismatch error name in .good

Signed-off-by: Anna Rift <anna.rift@hpe.com>
